### DISCUSSION TITLE
请问如何在yolov8中设置类别权重class_weights

### DISCUSSION THREAD

**Discussion by YWJ-sudo** [ultralytics/ultralytics#21690] [Q&A]
我曾在Ultralytics的讨论区看到可以在data.yaml文件中设置class_weights:[]的方法来为不同的类别设置不同的类别权重，但是在训练过程中，cls_loss几乎没有任何区别，我难以确定这样的设置方法是否生效
<img width="1505" height="777" alt="屏幕截图 2025-08-14 165620" src="https://github.com/user-attachments/assets/55e11d16-ca01-4b83-96a6-7a7953828d39" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @YWJ-sudo, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get started quickly—an Ultralytics engineer will also assist you here soon.

We recommend a visit to the Docs for guidance and examples, where many common questions may already be answered:
- Python: https://docs.ultralytics.com/usage/python/
- CLI: https://docs.ultralytics.com/usage/cli/

If this is a 🐛 Bug Report, please provide a minimum reproducible example (MRE) to help us debug it:
- https://docs.ultralytics.com/help/minimum-reproducible-example/

If this is a custom training ❓ Question, please provide as much information as possible, including:
- Training command and exact Ultralytics version
- Relevant config snippets (e.g., data.yaml), training logs, and dataset samples
- Confirmation you’re following our Tips for Best Training Results: https://docs.ultralytics.com/guides/model-training-tips/

Join the Ultralytics community where it suits you best:
- Discord for real-time chat 🎧: https://discord.com/invite/ultralytics
- Discourse for in-depth discussions: https://community.ultralytics.com/
- Subreddit for community threads: https://www.reddit.com/r/Ultralytics/

## Upgrade

Please upgrade to the latest ultralytics package including all requirements in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

Requirements: https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- Google Cloud Deep Learning VM. See GCP Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/
- Amazon Deep Learning AMI. See AWS Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/
- Docker Image. See Docker Quickstart Guide: https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/ <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all Ultralytics CI tests are currently passing. CI tests verify correct operation of all YOLO Modes and Tasks on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
感谢反馈。当前 Ultralytics YOLO（含 v8/YOLO11）训练并不会从 data.yaml 读取 class_weights 并应用到分类损失，cls_loss 使用未加权的 BCEWithLogitsLoss 计算，见 v8DetectionLoss 分类损失实现文档中的分类项说明。若需缓解类不平衡，可提高训练参数文档中的 cls 超参或在数据层进行重采样；若必须按类加权，需自定义 ultralytics/utils/loss.py，在 BCE 的 pos_weight 或 target_scores 上引入权重。若你看到有文档称支持 data.yaml: class_weights，请贴出链接，我们会核实并修正。

**Comment by Y-T-G** [comment]
You can try this for class balancing:

https://y-t-g.github.io/tutorials/yolo-class-balancing/

**Comment by yyq19990828** [comment]
我之前看到过，源码里的类权重还在注释(TODO)阶段，如果你想解决类不平衡的话，可以尝试给检测头中的类输出分支，初始化不同的bias

**Comment by YWJ-sudo** [comment]
---原始邮件---
发件人: "Mohammed ***@***.***&gt;
发送时间: 2025年8月16日(周六) 上午7:09
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [ultralytics/ultralytics] 请问如何在yolov8中设置类别权重class_weights (Discussion #21690)




 
You can try this for class balancing:
 
https://y-t-g.github.io/tutorials/yolo-class-balancing/

**Comment by YWJ-sudo** [comment]
初始化不同的偏置?是指数量较少的类赋予较大的偏置吗？可以劳烦你详细和我说说如何实现吗

---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2025年8月17日(周日) 晚上7:39
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [ultralytics/ultralytics] 请问如何在yolov8中设置类别权重class_weights (Discussion #21690)




 
我之前看到过，源码里的类权重还在注释(TODO)阶段，如果你想解决类不平衡的话，可以尝试给检测头中的类输出分支，初始化不同的bias

**Comment by YWJ-sudo** [comment]
感谢你的回复，这对我非常有帮助，非常感谢！

---原始邮件---
发件人: "Mohammed ***@***.***&gt;
发送时间: 2025年8月16日(周六) 上午7:09
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [ultralytics/ultralytics] 请问如何在yolov8中设置类别权重class_weights (Discussion #21690)




 
You can try this for class balancing:
 
https://y-t-g.github.io/tutorials/yolo-class-balancing/

**Comment by yyq19990828** [comment]
有一套逻辑的 我不是很熟 详见[苏剑林博客](https://kexue.fm/search/%E7%B1%BB%E4%B8%8D%E5%B9%B3%E8%A1%A1/)

**Comment by YWJ-sudo** [comment]
虽然我只看了相当少的一部分内容，但也觉得对我有莫大的帮助，非常感谢你的回复及分享



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2025年8月17日(周日) 晚上8:19
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [ultralytics/ultralytics] 请问如何在yolov8中设置类别权重class_weights (Discussion #21690)




 
有一套逻辑的 我不是很熟 详见苏剑林博客

**Reply by glenn-jocher** [reply]
是的，思路是给稀有类更大的初始正例先验，把检测头的分类分支 bias 按类设置为 logit(先验概率)；实现上直接在训练前修改 Detect 头各层的 cls 卷积 bias 即可，公式见 bias_init_with_prob 文档和检测头 bias 初始化参考文档。示例（v8/YOLO11通用，放在 train 前执行）:

```python
from ultralytics import YOLO
import torch

model = YOLO('yolo11n.yaml')  # 也可加载你的v8/pt权重
head = model.model.model[-1]  # Detect() 头
nc = head.nc
# 按类设置先验概率(0,1)，稀有类可设大一些如0.05~0.1；未知先用0.01基线
priors = torch.full((nc,), 0.01)
bias = torch.log(priors / (1 - priors))  # 等价于 bias_init_with_prob 的逐类形式

for cls_branch in head.cv3:  # 分类分支
    cls_branch[-1].bias.data.copy_(bias.to(cls_branch[-1].bias.device))

model.train(data='data.yaml', model='yolo11n.yaml', epochs=...)
```

参考：bias_init_with_prob 和 Initialize Detect() biases。注意这只影响初始化，训练若干迭代后会被学习覆盖，用于缓解早期不平衡；请继续在此帖交流，我们不通过邮件提供私下支持。
